src/shaders/pbr.frag

```diff
@@ -0,0 +1,292 @@
+#if defined(OBJECT_ID)
+layout(location = OUTPUT_ATTRIBUTE_LOCATION_OBJECT_ID) out highp uint
+    fragmentObjectId;
+#endif
+
+#if (LIGHT_COUNT > 0)
+// -------------- lights -------------------
+// NOTE: In this shader, the light intensity is considered in the lightColor!!
+uniform vec3 LightColors[LIGHT_COUNT];
+uniform float LightRanges[LIGHT_COUNT];
+
+// if .w == 0, it means it is a dirctional light, .xyz is the direction;
+// if .w == 1, it means it is a point light, .xyz is the light position;
+// it is NOT put in the Light Structure, simply because we may modify the code
+// so it is computed in the vertex shader.
+uniform vec4 LightDirections[LIGHT_COUNT];
+#endif
+
+// -------------- material, textures ------------------
+struct MaterialData {
+  vec4 baseColor;     // diffuse color, if baseColorTexture exists,
+                      // multiply it with the baseColorTexture
+  float roughness;    // roughness of a surface, if roughness texture exists,
+                      // multiply it with the roughnessTexture
+  float metallic;     // metalness of a surface, if metallic texture exists,
+                      // multiply it the metallicTexture
+  vec3 emissiveColor; // emissiveColor, if emissive texture exists,
+                      // multiply it the emissiveTexture
+};
+uniform MaterialData Material;
+
+#if defined(BASECOLOR_TEXTURE)
+uniform sampler2D baseColorTexture;
+#endif
+#if defined(ROUGHNESS_TEXTURE)
+uniform sampler2D roughnessTexture;
+#endif
+#if defined(METALLIC_TEXTURE)
+uniform sampler2D metallicTexture;
+#endif
+#if defined(NORMAL_TEXTURE)
+uniform sampler2D normalTexture;
+#endif
+
+#if defined(NONE_ROUGHNESS_METALLIC_TEXTURE) || defined(OCCLUSION_ROUGHNESS_METALLIC_TEXTURE)
+uniform sampler2D packedTexture;
+#endif
+
+#if defined(EMISSIVE_TEXTURE)
+uniform sampler2D emissiveTexture;
+#endif
+
+// -------------- uniforms ----------------
+#if defined(OBJECT_ID)
+uniform highp uint objectId;
+#endif
+
+#if defined(NORMAL_TEXTURE) && defined(NORMAL_TEXTURE_SCALE)
+uniform mediump float normalTextureScale
+#ifndef GL_ES
+    = 1.0
+#endif
+    ;
+#endif
+
+// -------------- shader ------------------
+#if defined(NORMAL_TEXTURE)
+vec3 getNormalFromNormalMap() {
+  vec3 tangentNormal =
+#if defined(NORMAL_TEXTURE_SCALE)
+      normalize((texture(normalTexture, texCoord).xyz * 2.0 - 1.0) *
+                vec3(normalTextureScale, normalTextureScale, 1.0));
+#else
+      texture(normalTexture, texCoord).xyz * 2.0 - 1.0;
+#endif
+
+#if defined(PRECOMPUTED_TANGENT)
+  mat3 TBN = mat3(tangent, biTangent, normal);
+#else
+  // Normal Mapping Without Precomputed Tangents
+  // See here: http://www.thetenthplanet.de/archives/1180
+  vec3 q1 = dFdx(position);
+  vec3 q2 = dFdy(position);
+  vec2 st1 = dFdx(texCoord);
+  vec2 st2 = dFdy(texCoord);
+
+  vec3 N = normalize(normal);
+  vec3 T = normalize(q1 * st2.t - q2 * st1.t);
+  vec3 B = -normalize(cross(N, T));
+  mat3 TBN = mat3(T, B, N);
```

Comment:
Didn't we agree we'd drop this and render without a normal map instead if there are no tangents (like done with `GenericDrawable` in #837) ? Since, as you said, this caused some weird artifacts in some cases.

Reply:
@mosra : Yes, we had that agreement.

But later I thought it might be good to leave it here as a fallback. It was correct in most cases.
I saw many systems using this trick, for example, here: 
https://github.com/SaschaWillems/Vulkan-glTF-PBR/blob/master/data/shaders/pbr_khr.frag#L125

Yes, true, there are artifacts in some corner cases. If you still feel uncomfortable, I can remove it though.

Reply:
It's easy to add tangents in an offline processing step (via Blender, for example) and most models already have them. The only PBR models that *didn't* have those were the ones exported from Blender without the Tangents option checked. 

This additional code creates another shader variant we need to test and maintain, so I'd vote for removing it -- the best code is code that's not written :P